File: T-Bank/Algorithms-and-Data-Structures-2025/13sprint-Mathematics/contest/F-Забывчивый_Матвей/solution.py
import sys


def power(x: int, y: int, p: int) -> int:
    output: int = 1
    x = x % p
    while y > 0:
        if y & 1: output = (output * x) % p
        y = y >> 1
        x = (x * x) % p
    return output;


def mod_inverse(n: int, mod: int) -> int:
    return power(n, mod - 2, mod);


def solution() -> None:
    """
    Time Complexity O(logN)
    Memory Complexity O(1)
    """
    n, m, k, MOD = map(int, sys.stdin.readline().rstrip().split())
    m_pow_n: int = power(m, n, MOD)
    k_inverse: int = mod_inverse(k, MOD)
    final_time: int = (m_pow_n * k_inverse) % MOD
    sys.stdout.write('{}\n'.format(final_time))
    

if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/13sprint-Mathematics/contest/F-Забывчивый_Матвей/solution.cpp
#include <iostream>


long long power(long long base, long long exp, long long mod) {
    long long output = 1;
    base %= mod;
    while (exp > 0) {
        if (exp % 2 == 1) output = (output * base) % mod;
        base = (base * base) % mod;
        exp /= 2;
    }
    return output;
}


long long modInverse(long long n, long long mod) {
    return power(n, mod - 2, mod);
}


void solution() {
    /*
    Time Complexity O(logN)
    Memory Complexity O(1)
    */
    long long N, M, K, MOD;
    std::scanf("%lld %lld %lld %lld", &N, &M, &K, &MOD);
    long long mPowN = power(M, N, MOD);
    long long kInverse = modInverse(K, MOD);
    long long finalTime = (mPowN * kInverse) % MOD;
    std::printf("%lld\n", finalTime);
}


int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);
    solution();
}